platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java

```diff
@@ -0,0 +1,210 @@
+import javax.annotation.Nullable;
+import javax.ws.rs.core.Response.StatusType;
+import org.apache.commons.lang.StringUtils;
+import org.apache.felix.utils.properties.Properties;
+import org.codice.ddf.cxf.client.ClientFactoryFactory;
+import org.codice.ddf.cxf.client.SecureCxfClientFactory;
+import org.codice.ddf.platform.util.uuidgenerator.UuidGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SolrPasswordUpdateImpl implements SolrPasswordUpdate {
+
+  private static final String ENC_TEMPLATE = "ENC(%s)";
+
+  private static final String SYSTEM_PROPERTIES_FILE = "custom.system.properties";
+  private static final String KARAF_ETC = "karaf.etc";
+  private static final String SET_USER_JSON_TEMPLATE = "{ \"set-user\": {\"%s\" : \"%s\"}}";
+  private static final Logger LOGGER = LoggerFactory.getLogger(SolrPasswordUpdateImpl.class);
+
+  @SuppressWarnings("squid:S2068" /* This constant does hold an actual password */)
+  private static final String SOLR_PASSWORD_PROPERTY_NAME = "solr.password";
+
+  private final ClientFactoryFactory restClientFactoryFactory;
+  private final UuidGenerator uuidGenerator;
+  private java.util.Properties properties;
+  private SolrAuthResource solrAuthResource;
+  private String newPasswordPlainText;
+  private String newPasswordEncrypted;
+  private EncryptionService encryptionService;
+  private StatusType solrResponse;
+  private String newPasswordWrappedEncrypted;
+
+  public SolrPasswordUpdateImpl(
+      UuidGenerator uuidGenerator,
+      ClientFactoryFactory clientFactoryFactory,
+      EncryptionService encryptionService) {
+    this.encryptionService = encryptionService;
+    this.uuidGenerator = uuidGenerator;
+    this.restClientFactoryFactory = clientFactoryFactory;
+    properties =
+        AccessController.doPrivileged(
+            (PrivilegedAction<java.util.Properties>) System::getProperties);
+  }
+
+  private void initialize() {
+    SecureCxfClientFactory<SolrAuthResource> factory =
+        restClientFactoryFactory.getSecureCxfClientFactory(
+            properties.getProperty("solr.http.url"),
+            SolrAuthResource.class,
+            properties.getProperty("solr.username"),
+            getPlaintextPasswordFromProperties());
+    solrAuthResource = factory.getClient();
+  }
+
+  /**
+   * If the password is not changed, change it. Only call this method if the system is configured to
+   * use basic authentication for Solr. This method will only change the password if it detects that
+   * the default password is still in use. This object is intended to be a singleton, a blueprint
+   * bean. That is why is can use itself as the lock
+   */
+  public synchronized void updateSolrPassword() {
+    if (configuredToAttemptAutoPasswordChange() && isUsingDefaultPassword()) {
+      initialize();
+      generatePassword();
+      setPasswordInSolr();
+      if (isSolrPasswordChangeSuccessfull()) {
+        setPasswordInMemory();
```

Comment:
❓ What happens if saving to a file fails? 

Memory will have been updated making it work for now and upon restart, the system wouldn't work anymore ...

Reply:
The only issue I have here (as opposed to above) is that the restart could happen many weeks from the time the issue occurred and those logs might not be available. And even if they were, it would be almost impossible to backtrack what happened since it could be weeks ... no way to correlate.

Reply:
What do you suggest?

Reply:
only updated memory if you were able to save to disk successfully?

Reply:
That seems reasonable.